pkg/registry/core/pod/storage/eviction.go

```diff
@@ -307,12 +310,25 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
 	return &metav1.Status{Status: metav1.StatusSuccess}, nil
 }
 
-func addConditionAndDeletePod(r *EvictionREST, ctx context.Context, name string, validation rest.ValidateObjectFunc, options *metav1.DeleteOptions) error {
-	if feature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) {
-		pod, err := getPod(r, ctx, name)
-		if err != nil {
-			return err
+func addConditionAndDeletePod(r *EvictionREST, ctx context.Context, name string, pod *api.Pod, deleteOptions *metav1.DeleteOptions) error {
+	if !dryrun.IsDryRun(deleteOptions.DryRun) && feature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) {
+		setPreconditions := func(_ context.Context, _, oldObj runtime.Object) (runtime.Object, error) {
+			// Throwaway the newObj. We care only about the latest pod obtained from etcd (oldObj).
+			// So we can add DisruptionTarget condition in conditionAppender without conflicts.
+			oldPodCopy := oldObj.(*api.Pod).DeepCopy()
+			if deleteOptions.Preconditions != nil {
+				// Insert preconditions into the pod object if we or the user specifically requested them,
+				// and let the update fail in case they do not pass.
+				if uid := deleteOptions.Preconditions.UID; uid != nil && len(*uid) > 0 {
```

Comment:
I'm wondering if the check `len(*uid) > 0` is needed. Shouldn't "" be just treated as an invalid value at the validation level and rejected?

Reply:
it seems we have only these validations for DeleteOptions:

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation.go#L116

which does not include Preconditions. 

@liggitt is it reasonable and can we make the validation stricter?

Reply:
no, don't change the validation there

when deleting and sending delete options via URL query parameter, preconditions are only created if greater than zero length:
```go
	if len(uid) > 0 || len(resourceVersion) > 0 {
		if out.Preconditions == nil {
			out.Preconditions = &Preconditions{}
		}
		if len(uid) > 0 {
			out.Preconditions.UID = &uid
		}
		if len(resourceVersion) > 0 {
			out.Preconditions.ResourceVersion = &resourceVersion
		}
	}
```

When sending delete options in the body of a delete request, you could set resourceVersion or UID to non-nil zero-length strings, though that would be a weird thing to do. Setting zero-length UID or resourceVersion strings in a pod while updating it will not have the same effect as those zero-length preconditions.

The UID would get copied from the existing pod
* https://github.com/kubernetes/kubernetes/blob/1cb334960c4288fc7830a16c97f5f8489f140b5b/staging/src/k8s.io/apiserver/pkg/registry/rest/update.go#L131-L134

Setting an empty resourceVersion in the pod would circumvent optimistic locking and mean we would stomp whatever was in etcd if we raced with another write
* https://github.com/kubernetes/kubernetes/blob/1cb334960c4288fc7830a16c97f5f8489f140b5b/pkg/registry/core/pod/strategy.go#L154-L157
https://github.com/kubernetes/kubernetes/blob/1cb334960c4288fc7830a16c97f5f8489f140b5b/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L576-L584

Reply:
keep the checks to only set UID / resourceVersion in the pod if they are non-empty

Reply:
ack, thanks for the explanation